tensorflow/contrib/rnn/python/ops/rnn_cell.py

```diff
@@ -3392,3 +3392,252 @@ def call(self, inputs, state):
+    https://arxiv.org/abs/1807.08518
+    Mark Collier, Joeran Beel
+
+    which is in turn based on the source code of:
+    https://github.com/snowkylin/ntm
+
+    and of course the original NTM paper:
+    Neural Turing Machines
+    https://arxiv.org/abs/1410.5401
+    A Graves, G Wayne, I Danihelka
+  """
+  def __init__(self, controller, memory_size,
+    memory_vector_dim, read_head_num, write_head_num, shift_range=1,
+    output_dim=None, clip_value=20, dtype=dtypes.float32,
+    reuse=None, name=None):
+    """Initialize the NTM Cell.
+
+      Args:
+        controller: an RNNCell, the RNN controller.
+        memory_size: int, The number of memory locations in the
+          NTM memory matrix
+        memory_vector_dim: int, The dimensionality of each location in the
+          NTM memory matrix
+        read_head_num: int, The number of read heads from the controller
+          into memory
+        write_head_num: int, The number of write heads from the controller
+          into memory
+        shift_range: int, The number of places to the left/right it is possible
+          to iterate the previous address to in a single step
+        output_dim: int, The number of dimensions to make a linear projection
+          of the NTM controller outputs to.
+          If None, no linear projection is applied
+        clip_value: float, The maximum absolute value the controller parameters
+          are clipped to
+        dtype: Default dtype of the layer (default of `None` means use the type
+          of the first input). Required when `build` is called before `call`.
+        reuse: (optional) Python boolean describing whether to reuse variables
+          in an existing scope.  If not `True`, and the existing scope already
+          has the given variables, an error is raised.
+        name: String, the name of the layer. Layers with the same name will
+          share weights, but to avoid mistakes we require reuse=True in such
+          cases.
+    """
+    super(NTMCell, self).__init__(_reuse=reuse, dtype=dtype, name=name)
+
+    rnn_cell_impl.assert_like_rnncell("NTM RNN controller cell", controller)
+
+    self.controller = controller
+    self.memory_size = memory_size
+    self.memory_vector_dim = memory_vector_dim
+    self.read_head_num = read_head_num
+    self.write_head_num = write_head_num
+    self.clip_value = clip_value
+    self.reuse = reuse
+
+    self.NTMControllerState = collections.namedtuple('NTMControllerState',
+      ('controller_state', 'read_vector_list', 'w_list', 'M'))
+
+    self.output_dim = output_dim
+    self.shift_range = shift_range
+
+    self.num_parameters_per_head = self.memory_vector_dim + \
+      2 * self.shift_range + 4
+    self.num_heads = self.read_head_num + self.write_head_num
+    self.total_parameter_num = self.num_parameters_per_head * self.num_heads + \
+      self.memory_vector_dim * 2 * self.write_head_num
+
+  @property
+  def state_size(self):
+    return self.NTMControllerState(
```

Comment:
ah i see; yeah just move NTMControllerState outside of the class and into the module.